<a href="https://colab.research.google.com/github/isabellamwright/project_voting/blob/main/PredictiveAlgorithm.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
! git clone https://github.com/isabellamwright/project_voting

Cloning into 'project_voting'...
remote: Enumerating objects: 117, done.
remote: Counting objects: 100% (33/33), done.
remote: Compressing objects: 100% (27/27), done.
remote: Total 117 (delta 17), reused 10 (delta 6), pack-reused 84
Receiving objects: 100% (117/117), 62.33 MiB | 6.68 MiB/s, done.
Resolving deltas: 100% (42/42), done.
Updating files: 100% (48/48), done.


In [2]:
import pandas as pd

# Load merged dataset
data = pd.read_csv("/content/project_voting/final.csv")

# Handle missing values if any
data.fillna(0, inplace=True)  # For simplicity, filling missing values with 0

# Explore the dataset
print(data.head())



   Unnamed: 0 Party       County Name     1 GIS Join Match Code  \
0           0     R   Accomack County  2823            G5100010   
1           1     D  Albemarle County  2824            G5100030   
2           2     R  Alleghany County  2825            G5100050   
3           3     R     Amelia County  2826            G5100070   
4           4     R    Amherst County  2827            G5100090   

  Data File Year State Postal Abbreviation  Region Code  Division Code  \
0      2016-2020                        VA          0.0            0.0   
1      2016-2020                        VA          0.0            0.0   
2      2016-2020                        VA          0.0            0.0   
3      2016-2020                        VA          0.0            0.0   
4      2016-2020                        VA          0.0            0.0   

  State Name  ...  Female: 35 to 64 years  \
0   Virginia  ...                    6474   
1   Virginia  ...                   20763   
2   Virginia  ...

In [4]:
print(data.columns)

Index(['Unnamed: 0', 'Party', 'County Name', '1', 'GIS Join Match Code',
       'Data File Year', 'State Postal Abbreviation', 'Region Code',
       'Division Code', 'State Name',
       ...
       'Female: 35 to 64 years', 'Female: 35 to 64 years: With a disability',
       'Female: 35 to 64 years: No disability', 'Female: 65 to 74 years',
       'Female: 65 to 74 years: With a disability',
       'Female: 65 to 74 years: No disability', 'Female: 75 years and over',
       'Female: 75 years and over: With a disability',
       'Female: 75 years and over: No disability', 'Gini Index'],
      dtype='object', length=164)


In [31]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
import pandas as pd

# Read the CSV file
df = pd.read_csv('/content/project_voting/final.csv')

# Select essential numeric features (you may adjust this based on your domain knowledge)
numeric_features = ['Gini Index', 'Total']

# Subset the data to include only essential numeric features and the target variable
data = df[numeric_features + ['Party']]

# Drop rows with missing values
data.dropna(inplace=True)

# Separate the target variable ('Party') from the features
y = data['Party'].apply(lambda x: 1 if x == 'D' else 0)
X = data.drop(['Party'], axis=1)

# Initialize the model
cart = DecisionTreeClassifier(random_state=42)

# Fit the model
cart.fit(X, y)

# Make predictions
y_pred = cart.predict(X)

# Calculate accuracy
accuracy = accuracy_score(y, y_pred)
print("Accuracy:", accuracy)


Accuracy: 1.0


<ipython-input-31-785c7aa25398>:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.dropna(inplace=True)


An accuracy of 1 means the decision tree model achieved perfect accuracy on the training data, which could indicate overfitting. The model has learned the training data too well and may not generalize well to unseen data. Now let's evaluate the model on a separate test dataset by splitting the data into training and testing sets.

In [32]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and fit the model
cart = DecisionTreeClassifier(random_state=42)
cart.fit(X_train, y_train)

# Make predictions on the test set
y_pred_test = cart.predict(X_test)

# Calculate accuracy on the test set
accuracy_test = accuracy_score(y_test, y_pred_test)
print("Accuracy on Test Set:", accuracy_test)


Accuracy on Test Set: 0.5925925925925926


An accuracy of .59259 indicates that the model's performance on the unseen data is lower than on the training data. This supports the argument above indicating that the model might be overfitting to some extent.

In [34]:
# Print column names in the DataFrame
print(df.columns)


Index(['Unnamed: 0', 'Party', 'County Name', '1', 'GIS Join Match Code',
       'Data File Year', 'State Postal Abbreviation', 'Region Code',
       'Division Code', 'State Name',
       ...
       'Female: 35 to 64 years', 'Female: 35 to 64 years: With a disability',
       'Female: 35 to 64 years: No disability', 'Female: 65 to 74 years',
       'Female: 65 to 74 years: With a disability',
       'Female: 65 to 74 years: No disability', 'Female: 75 years and over',
       'Female: 75 years and over: With a disability',
       'Female: 75 years and over: No disability', 'Gini Index'],
      dtype='object', length=164)


In [36]:
# Print all column names in the DataFrame
for column in df.columns:
    print(column)


Unnamed: 0
Party
County Name
1
GIS Join Match Code
Data File Year
State Postal Abbreviation
Region Code
Division Code
State Name
State Code
County Code
County Subdivision Code
Place Code
Census Tract Code
Consolidated City Code
American Indian Area/Alaska Native Area/Hawaiian Home Land Code
American Indian Area/Alaska Native Area (Reservation or Statistical Entity Only) Code
American Indian Area (Off-Reservation Trust Land Only)/Hawaiian Home Land Code
American Indian Trust Land/ Hawaiian Home Land Indicator
Tribal Subdivision/Remainder Code
Alaska Native Regional Corporation Code
Metropolitan Statistical Area/Micropolitan Statistical Area Code
Combined Statistical Area Code
Metropolitan Division Code
New England City and Town Area Code
Combined New England City and Town Area Code
New England City and Town Area Division Code
Urban Area Code
Congressional District (2019-2021, 116th Congress) Code
State Legislative District (Upper Chamber) Code
State Legislative District (Lower Chamber) 

In [38]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
import pandas as pd

# Read the CSV file
df = pd.read_csv('/content/project_voting/final.csv')

# Select relevant features
relevant_features = [
    'U.S. citizen, born in the United States',
    'Not a U.S. citizen',
    'Income in the past 12 months below poverty level',
    'Party'
]

# Subset the data to include only relevant features
data = df[relevant_features].copy()  # Make a copy of the DataFrame

# Drop rows with missing values
data.dropna(inplace=True)

# Map party affiliation to binary labels (D: 1, R: 0)
data['Party'] = data['Party'].map({'D': 1, 'R': 0})

# Separate the target variable ('Party') from the features
y = data['Party']
X = data.drop(['Party'], axis=1)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and fit the model
cart = DecisionTreeClassifier(random_state=42)
cart.fit(X_train, y_train)

# Make predictions on the test set
y_pred_test = cart.predict(X_test)

# Calculate accuracy on the test set
accuracy_test = accuracy_score(y_test, y_pred_test)
print("Accuracy on Test Set:", accuracy_test)



Accuracy on Test Set: 0.8518518518518519


Accuracy in the context of this project represents the proportion of correctly predicted outcomes (party affiliations) by the model on the test set.

The accuracy of approximately 85% means that the model correctly predicted the party affiliation (D for Democratic or R for Republican) for 85% of the counties in the test set.

It also implies that the model's predictions align with the actual party affiliations of the counties in the test set about 85% of the time.

Interpreting accuracy in this way gives an understanding of how well the model performs in terms of classifying counties into their respective political parties based on the features provided in the dataset.

By incorporating additional features such as U.S. citizenship status, birthplace, and income level into the model, we observed a substantial improvement in predictive performance. Specifically, our model achieved an accuracy of approximately 85% on the test set, which represents a significant enhancement compared to the original model's accuracy of 59%.

This improvement underscores the relevance and predictive power of these features in determining party affiliation within Virginia counties. Factors such as citizenship status and income level likely play a crucial role in shaping political preferences and voting behaviors at the county level. By leveraging this demographic and socioeconomic information, our model can more accurately discern patterns and trends in party affiliation across different regions.

Now proceeding with other relevant columns to test our model with.

I want to add the following relevant varibles but it's not working!

In [46]:
# Select relevant columns for feature engineering
selected_columns = ['Income in the past 12 months at or above poverty level',
                    'Income in the past 12 months at or above poverty level: Male',
                    'Income in the past 12 months at or above poverty level: Female',
                    'Male',
                    'Female',
                    'Party']  # Include the target variable

# Create a new DataFrame with selected columns
selected_data = data[selected_columns].copy()

# Check for missing values
print("Missing values before preprocessing:")
print(selected_data.isnull().sum())

# Handle missing values (if any)
# For simplicity, let's drop rows with missing values in this example
selected_data.dropna(inplace=True)

# Split the data into features (X) and target (y)
X = selected_data.drop('Party', axis=1)
y = selected_data['Party']

# Display the first few rows of the preprocessed data
print("Preprocessed data:")
print(X.head())
print("Target variable (Party):")
print(y.head())


KeyError: "['Income in the past 12 months at or above poverty level', 'Income in the past 12 months at or above poverty level: Male', 'Income in the past 12 months at or above poverty level: Female', 'Male', 'Female'] not in index"